userspace/libsinsp/dns_manager.cpp

```diff
@@ -20,223 +20,346 @@ limitations under the License.
 {
 public:
 	typedef uint32_t type;
-	static type value(void *addr){ return *(uint32_t *)addr; }
+	static type value(void* addr) { return *(uint32_t*)addr; }
+	static std::string to_string(const type& v) { return ip2str(AF_INET, &v); }
 };
 
 // ipv6
-template <> class af_converter<AF_INET6>
+template<>
+class af_converter<AF_INET6>
 {
 public:
 	typedef std::pair<uint64_t, uint64_t> type;
-	static type value(void *addr)
+	typedef typename tbb::tbb_hash_compare<ipv6addr> hash_t;
+
+	static type value(void* addr)
 	{
-		auto * a = (uint64_t*)(addr);
+		auto* a = (uint64_t*)(addr);
 		return {a[0], a[1]};
 	}
 
-	typedef typename tbb::tbb_hash_compare<ipv6addr> hash_t;
+	static std::string to_string(const type& v)
+	{
+		uint64_t val[2];
+		val[0] = v.first;
+		val[1] = v.second;
+		return ip2str(AF_INET6, val);
+	}
 };
 
-// dns info to hold timestamps and resolution results.
-// results are intermediate to be inserted into map containers
-class dns_info
+template<int AF>
+class container
 {
-private:
+	using addr_t = typename af_converter<AF>::type;
 
-	uint64_t m_max_refresh_timeout = 0;
-	uint64_t m_refresh_timeout = 0;
-	uint64_t m_last_resolve_ts = 0;
-
-	template <int AF>
-	class container
+public:
+	bool has_address(void* addr, uint64_t ts)
 	{
-	private:
-		std::set<typename af_converter<AF>::type> m_addrs;
-	public:
-		bool has_address(void* addr) const
+		const auto it = m_addrs.find(af_converter<AF>::value(addr));
+		const auto ret = it != m_addrs.end();
+
+		if (ret)
 		{
-			return m_addrs.find(af_converter<AF>::value(addr)) != m_addrs.end();
+			it->second = ts;
 		}
 
-		void insert (void* addr)
+		G_LOG_FORMAT(sinsp_logger::SEV_DEBUG,
+		             "dns_info: lookup address: %s in: [%s]; found=%s",
+		             ip2str(AF, addr).c_str(),
+		             to_string().c_str(),
+		             ret ? "true" : "false");
+		return ret;
+	}
+
+	bool insert(void* addr, uint64_t ts)
+	{
+		const auto it = m_addrs.find(af_converter<AF>::value(addr));
+		if (it == m_addrs.end())
 		{
-			m_addrs.insert(af_converter<AF>::value(addr));
+			// new address
+			m_addrs.template emplace(af_converter<AF>::value(addr), ts);
+			return true;
 		}
 
-		template <template <typename T> class I>
-		void copy (I<typename af_converter<AF>::type>& inserter) const
+		// already resolved
+		it->second = ts;
+		return false;
+	}
+
+	void gc(uint64_t erase_timeout, uint64_t ts)
+	{
+		auto n_erased = 0;
+		for (auto it = m_addrs.begin(); it != m_addrs.end();)
 		{
-			if (!m_addrs.empty())
+			if (it->second + erase_timeout < ts)
+			{
+				++n_erased;
+				it = m_addrs.erase(it);
+			}
+			else
 			{
-				inserter.insert(m_addrs.begin(), m_addrs.end());
+				++it;
 			}
 		}
 
-		bool operator !=(const container& other)
+		if (n_erased > 0)
 		{
-			return m_addrs != other.m_addrs;
+			G_LOG_FORMAT(sinsp_logger::SEV_DEBUG,
+			             "dns_info gc: removed: %d out of: %d",
+			             n_erased,
+			             (int32_t)m_addrs.size());
```

Comment:
This is slightly misleading: "out of: %d" will show the number of items remaining, not the original count

```suggestion
			G_LOG_FORMAT(sinsp_logger::SEV_DEBUG,
			             "dns_info gc: removed: %d remaining: %d",
			             n_erased,
			             (int32_t)m_addrs.size());
```